src/Orleans.Core/Diagnostics/ActivityPropagationGrainCallFilter.cs

```diff
@@ -0,0 +1,94 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading.Tasks;
+
+namespace Orleans.Runtime
+{
+    internal static class ActivityPropagationGrainCallFilter
+    {
+        private const string TraceParentHeaderName = "traceparent";
+        private const string TraceStateHeaderName = "tracestate";
+
+        internal const string ActivitySourceName = "orleans.runtime.graincall";
+        internal const string ActivityNameIn = "Orleans.Runtime.GrainCall.In";
+        internal const string ActivityNameOut = "Orleans.Runtime.GrainCall.Out";
+
+        private static readonly ActivitySource activitySource = new(ActivitySourceName);
+
+        private static async Task ProcessNewActivity(IGrainCallContext context, string activityName, ActivityKind activityKind, ActivityContext activityContext)
+        {
+            // rpc attributes from https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/rpc.md
+            ActivityTagsCollection tags = null;
+            if (activitySource.HasListeners())
+                tags = new ActivityTagsCollection
+                {
+                    {"rpc.service", context.InterfaceMethod?.DeclaringType},
+                    {"rpc.method", context.InterfaceMethod?.Name},
+                    {"net.peer.name", context.Grain},
+                };
+
+            using var activity = activitySource.StartActivity(activityName, activityKind, activityContext, tags);
+            if (activity is not null)
+                RequestContext.Set(TraceParentHeaderName, activity.Id);
+
+            try
+            {
+                await context.Invoke();
+            }
+            catch (Exception e)
+            {
+                if (activity is not null && activity.IsAllDataRequested)
+                {
+                    // exception attributes from https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/exceptions.md
+                    activity.SetTag("exception.type", e.GetType());
```

Comment:
I think you might want to specify the `Name` or `FullName` of the exception type instead of the `Type` object itself?